river/cluster/textclust.py

```diff
@@ -0,0 +1,643 @@
+    >>> from river import metrics
+    >>> from river import cluster
+
+    >>> corpus = [
+    ...    {"text":'This is the first document.',"idd":1, "cluster": 1, "cluster":1},
+    ...    {"text":'This document is the second document.',"idd":2,"cluster": 1},
+    ...    {"text":'And this is super unrelated.',"idd":3,"cluster": 2},
+    ...    {"text":'Is this the first document?',"idd":4,"cluster": 1},
+    ...    {"text":'This is super unrelated as well',"idd":5,"cluster": 2},
+    ...    {"text":'Test text',"idd":6,"cluster": 5}
+    ... ]
+
+    >>> stopwords = [ 'stop', 'the', 'to', 'and', 'a', 'in', 'it', 'is', 'I']
+
+    >>> metric = metrics.AdjustedRand()
+
+    >>> model = compose.Pipeline(
+    ...     feature_extraction.BagOfWords(lowercase=True, ngram_range=(1, 2), stop_words=stopwords),
+    ...     cluster.TextClust(real_time_fading=False, fading_factor=0.001, tgap=100, auto_r=True,
+    ...     radius=0.9)
+    ... )
+
+    >>> for x in corpus:
+    ...     y_pred = model.predict_one(x["text"])
+    ...     y = x["cluster"]
+    ...     metric = metric.update(y,y_pred)
+    ...     model = model.learn_one(x["text"], id=x["idd"])
+
+    >>> print(metric)
+    AdjustedRand: -0.17647058823529413
+
+
+    """
+
+    # constructor with default specification
+    def __init__(
+        self,
+        radius=0.3,
+        fading_factor=0.0005,
+        tgap=100,
+        term_fading=True,
+        real_time_fading=True,
+        micro_distance="tfidf_cosine_distance",
+        macro_distance="tfidf_cosine_distance",
+        num_macro=3,
+        min_weight=0,
+        auto_r=False,
+        auto_merge=True,
+        sigma=1,
+    ):
+
+        self.radius = radius
+        self.fading_factor = fading_factor
+        self.tgap = tgap
+        self.term_fading = term_fading
+        self.micro_distance = micro_distance
+        self.macro_distance = macro_distance
+        self.num_macro = num_macro
+        self.real_time_fading = real_time_fading
+        self.min_weight = min_weight
+        self.auto_r = auto_r
+        self.auto_merge = auto_merge
+        self.sigma = sigma
+
+        # Initialize important values
+
+        self.t = None
+        self.last_cleanup = 0
+        self.n = 1
+        self.omega = 2 ** (-1 * self.fading_factor * self.tgap)
+        self.assignment = dict()
+        self.micro_clusters = dict()
+        self.microToMacro = None
+        self.realtime = None
+
+        self._clusterId = 0
+        self._up_to_date = False
+        self._dist_mean = 0
+        self._num_merged_obs = 0
+
+        # create a new distance instance for micro and macro distances.
+        self.micro_distance = self.distances(self.micro_distance)
+        self.macro_distance = self.distances(self.macro_distance)
+
+    def learn_one(self, x, t=None, sample_weight=None, **kwargs):
+
+        ## if id is specified
+        if kwargs["id"]:
+            id = kwargs["id"]
+        else:
+            id = None
```

Comment:
You have to document this behavior somewhere for users to understand.

I think we should look into consistently having an `id` parameter in the `learn_one` method of every clustering method (cc @hoanganhngo610)

Reply:
@MaxHalford  I also agree, since CluStream also depends a lot on distinguishing different micro clusters with unique IDs. My main concern is that how do we choose the format for the ID when two micro-clusters are formed?

Reply:
I don't know, I'll have to look into it. Let's not mix sample IDs with cluster IDs, I'm not sure which ones you're talking about

Reply:
Yes, @MaxHalford is right. These are two different things. 

So why do we need IDs for observations? Especially for text-based input data, a lot of metadata is associated (thinking of Tweets that hold user objects, locations, and other important information). This information can be used within post hoc analyses to find additional characteristics of clusters that are only built based on the text (e.g., which users belong to a certain cluster). Therefore it is necessary to store the IDs. The question here is if IDs should also be subject to concept drift to understand the current importance of observations. What I think is super important is that we do not "fade away" cluster IDs. It should always be possible to know which ID belongs to which cluster. However, we also have to consider what happens if a cluster itself is removed from the current clustering (maybe cluster - ID assignment should be persisted in secondary memory?)

So why do we need IDs for clusters? To understand how clusters were formed over time (e.g., merged together, etc.)

Reply:
I understand that we need to identify samples with identifiers. But I'm still a bit doubtful as to why this would have to be handled by the clustering model. Can't we just store the cluster assignments outside the clustering model? Indeed, it sounds like a feature you would want to do for any clustering task.

> So why do we need IDs for clusters? To understand how clusters were formed over time (e.g., merged together, etc.)

That sounds like it should be handled internally.

Reply:
> Can't we just store the cluster assignments outside the clustering model?

Yes, we could do this. It would be up to the user to ensure storage. If we decide on this, we should provide some copy&paste examples as I think this is something that people want to do frequently.

Reply:
Yes it's a good excuse to document it properly. Something in the user guide would be nice. You can kill two birds with one stone when you work on the notebook in a future PR :).

I'll let you remove this part from the code, then we can merge and move on.